File: tests/cases/enforce/immutable/python/inputs/src/orders/receipt.py
from cdec_rules import immutable


@immutable
class Receipt:
    def __init__(self, total: float, lines: int) -> None:
        # Constructor assignment is initialisation, not mutation.
        self.total: float = total
        self.lines: int = lines

    def with_discount(self, pct: float) -> "Receipt":
        """A 'modification' that returns a new instance. No assignment to
        `self`, so nothing fires."""
        return Receipt(self.total * (1 - pct), self.lines)

    def formatted(self) -> str:
        """Reads only."""
        return f"{self.lines} line(s)"

    def apply_discount(self, pct: float) -> None:
        """The violation: reassigns a field outside `__init__`."""
        self.total *= 1 - pct

File: tests/cases/enforce/immutable/csharp/inputs/src/Orders/Receipt.cs
using CodeConstraints.Rules;

namespace Orders;

[Immutable]
public class Receipt
{
    private decimal _total;
    private int _lines;

    // Constructor assignment is initialisation, not mutation.
    public Receipt(decimal total, int lines)
    {
        _total = total;
        _lines = lines;
    }

    // Returns a new instance instead of mutating. Silent.
    public Receipt WithDiscount(decimal pct)
    {
        return new Receipt(_total * (1 - pct), _lines);
    }

    // Reads only. Silent.
    public string Formatted() { return $"{_lines} line(s)"; }

    // The violation: reassigns a field outside the constructor.
    public void ApplyDiscount(decimal pct)
    {
        _total = _total * (1 - pct);
    }
}
